x/oracle/keeper/prices.go

```diff
@@ -0,0 +1,189 @@
+func (k Keeper) SetPrices(ctx sdk.Context, prices types.Prices) {
+	store := k.getPriceTRStore(ctx, prices.TokenId)
+	for _, v := range prices.PriceList {
+		b := k.cdc.MustMarshal(v)
+		store.Set(types.PricesRoundKey(v.RoundId), b)
+	}
+	store.Set(types.PricesNextRountIdKey, types.Uint64Bytes(prices.NextRountId))
+}
+
+// GetPrices returns a prices from its index
+func (k Keeper) GetPrices(
+	ctx sdk.Context,
+	tokenId int32,
+
+) (val types.Prices, found bool) {
+	store := k.getPriceTRStore(ctx, tokenId)
+	//	nextRoundIdB := store.Get(types.PricesNextRountIdKey)
+	//	if nextRoundIdB == nil {
+	//		return val, false
+	//	}
+
+	//	nextRoundId := binary.BigEndian.Uint64(nextRoundIdB)
+	nextRoundId := k.GetNextRoundId(ctx, tokenId)
+
+	val.TokenId = tokenId
+	val.NextRountId = nextRoundId
+	val.PriceList = make([]*types.PriceWithTimeAndRound, nextRoundId)
+	//0 roundId is reserved, expect the roundid corresponds to the slice index
+	val.PriceList[0] = &types.PriceWithTimeAndRound{}
+	for i := uint64(1); i < nextRoundId; i++ {
+		b := store.Get(types.PricesRoundKey(i))
+		val.PriceList[i] = &types.PriceWithTimeAndRound{}
+		if b != nil {
+			//should alwyas be true since we don't delete prices from history round
+			k.cdc.MustUnmarshal(b, val.PriceList[i])
+			found = true
+		}
+	}
+
+	return
+}
+
+// RemovePrices removes a prices from the store
+func (k Keeper) RemovePrices(
+	ctx sdk.Context,
+	tokenId int32,
+
+) {
+	store := k.getPriceTRStore(ctx, tokenId)
+	//	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	iterator := store.Iterator(nil, nil)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		store.Delete(iterator.Key())
+	}
+}
+
+// GetAllPrices returns all prices
+func (k Keeper) GetAllPrices(ctx sdk.Context) (list []types.Prices) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.PricesKeyPrefix))
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+
+	defer iterator.Close()
+
+	//	prevTokenId := uint32(0)
+	//	var val types.PriceWithTimeAndRound
+	var price types.Prices
+	prevTokenId := int32(0)
+	for ; iterator.Valid(); iterator.Next() {
+		tokenId, _, nextRoundId := parseKey(iterator.Key())
+		if prevTokenId == 0 {
+			prevTokenId = tokenId
+			price.TokenId = tokenId
+		} else if prevTokenId != tokenId && price.TokenId > 0 {
+			list = append(list, price)
+			prevTokenId = tokenId
+			price = types.Prices{TokenId: tokenId}
+		}
+		if nextRoundId {
+			price.NextRountId = binary.BigEndian.Uint64(iterator.Value())
+		} else {
+			var val types.PriceWithTimeAndRound
+			k.cdc.Unmarshal(iterator.Value(), &val)
+			price.PriceList = append(price.PriceList, &val)
+		}
+	}
+	if price.TokenId > 0 {
+		list = append(list, price)
+	}
+	return
+}
+
+func (k Keeper) AppendPriceTR(ctx sdk.Context, tokenId int32, priceTR types.PriceWithTimeAndRound) {
+	nextRoundId := k.GetNextRoundId(ctx, tokenId)
+	if nextRoundId != priceTR.RoundId {
```

Comment:
Return an error instead of failing silently        
If the round ID of the price information you are trying to add is not the expected next round ID, the caller will receive a clear error message indicating the problem

Reply:
This method will be invoked in endblock. It won't respond to the caller directly, but we could certainly add some return info for the log.